examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,391 @@
+        seq_len = input_shape[1]
+        causal_mask = self.causal_attention_mask(batch_size, seq_len, seq_len, tf.bool)
+        trg_att = self.self_att(trg, trg, attention_mask=causal_mask)
+        trg_norm = self.ln1(trg + self.drop(trg_att))
+        enc_out = self.enc_att(trg_norm, enc_out)
+        enc_out_norm = self.ln2(self.drop(enc_out) + trg_norm)
+        ffn_out = self.ffn(enc_out_norm)
+        ffn_out_norm = self.ln3(enc_out_norm + self.drop(ffn_out))
+        return ffn_out_norm
+
+
+"""
+## Complete Transformer Model
+"""
+loss_object = tf.keras.losses.SparseCategoricalCrossentropy(
+    from_logits=True, reduction="none"
+)
+
+
+def masked_loss(real, pred):
+    """ assuming pad token index = 0 """
+    mask = tf.math.logical_not(tf.math.equal(real, 0))
+    loss_ = loss_object(real, pred)
+    mask = tf.cast(mask, dtype=loss_.dtype)
+    loss_ *= mask
+    return tf.reduce_mean(loss_)
+
+
+class Transformer(keras.Model):
+    def __init__(
+        self,
+        input_type="tokens",
+        nvocab=1000,
+        ninp=32,
+        nhid=64,
+        nhead=2,
+        nff=128,
+        src_maxlen=100,
+        trg_maxlen=100,
+        nlayers=2,
+        nclasses=10,
+    ):
+        super().__init__()
+        self.nlayers = nlayers
+        self.trg_maxlen = trg_maxlen
+        self.input_type = input_type
+
+        self.enc_input = TransformerInput(
+            input_type=input_type, nvocab=1000, nhid=64, maxlen=src_maxlen
+        )
+        self.dec_input = TransformerInput(
+            input_type="tokens", nvocab=nclasses, nhid=nhid, maxlen=trg_maxlen
+        )
+        for i in range(nlayers):
+            setattr(self, f"enc_layer_{i}", TransformerEncoderLayer(nhid, nhead, nff))
+            setattr(self, f"dec_layer_{i}", TransformerDecoderLayer(nhid, nhead, nff))
+
+        self.final = layers.Dense(nclasses)
+
+    def encode_src(self, src):
+        x = self.enc_input(src)
+        for i in range(self.nlayers):
+            x = getattr(self, f"enc_layer_{i}")(x)
+        return x
+
+    def decode(self, enc_out, trg):
+        y = self.dec_input(trg)
+        for i in range(self.nlayers):
+            y = getattr(self, f"dec_layer_{i}")(enc_out, y)
+        return y
+
+    def call(self, inputs):
+        src = inputs[0]
+        trg = inputs[1]
+        x = self.encode_src(src)
+        y = self.decode(x, trg)
+        return self.final(y)
+
+    def train_step(self, batch):
+        """ Process one batch inside model.fit() """
+        src = batch["src"]
+        trg = batch["trg"]
+        dec_inp = trg[:, :-1]
+        dec_trg = trg[:, 1:]
+        with tf.GradientTape() as tape:
+            preds = self([src, dec_inp])
+            loss = masked_loss(dec_trg, preds)
+        trainable_vars = self.trainable_variables
+        gradients = tape.gradient(loss, trainable_vars)
+        self.optimizer.apply_gradients(zip(gradients, trainable_vars))
+        return {"loss": loss}
```

Comment:
For loss tracking, use a `Mean` metric. Otherwise you're displaying the per-batch loss (high variance) instead of the running average of the loss over each epoch (more stable). Also create a `metrics` `@property` listing the tracker, so the model is able to reset the average after every epoch. See e.g. the GAN example.